org.adempiere.base/src/org/compiere/model/MOrder.java

```diff
@@ -1263,12 +1263,42 @@ protected boolean beforeSave (boolean newRecord)
 					return false;
 			}
 		}
-		
+
+		// IDEMPIERE-6046 - verify if the locations pertain to the BP
+		if (is_new() || is_ValueChanged(COLUMNNAME_C_BPartner_ID)) {
+			if (getC_BPartner_Location_ID() > 0) {
+				MBPartnerLocation bpl = new MBPartnerLocation(getCtx(), getC_BPartner_Location_ID(), get_TrxName());
+				if (bpl.getC_BPartner_ID() != getC_BPartner_ID()) {
+					set_ValueNoCheck(COLUMNNAME_C_BPartner_Location_ID, null);
+				}
+			}
+			if (getAD_User_ID() > 0) {
+				MUser user = new MUser(getCtx(), getAD_User_ID(), get_TrxName());
+				if (user.getC_BPartner_ID() != getC_BPartner_ID()) {
+					set_Value(COLUMNNAME_AD_User_ID, null);
+				}
+			}
+		}
+		if (is_new() || is_ValueChanged(COLUMNNAME_Bill_BPartner_ID)) {
+			if (getBill_Location_ID() > 0) {
+				MBPartnerLocation bpl = new MBPartnerLocation(getCtx(), getBill_Location_ID(), get_TrxName());
+				if (bpl.getC_BPartner_ID() != getBill_BPartner_ID()) {
+					set_Value(COLUMNNAME_Bill_Location_ID, null);
+				}
+			}
+			if (getBill_User_ID() > 0) {
+				MUser user = new MUser(getCtx(), getBill_User_ID(), get_TrxName());
+				if (user.getC_BPartner_ID() != getBill_BPartner_ID()) {
+					setBill_User_ID(-1);
+				}
+			}
+		}
+
```

Comment:
the validation looks trivial but I'm wondering the impact on performance

Reply:
Yes, it is by primary key but it is loading the whole PO although the validation just need C_BPartner_ID from AD_User or C_BPartner_Location.

Reply:
understood, changed the approach for direct SQL